### PR TITLE
Miscellaneous clean-up and minor fixes.

### LINKED ISSUE #404: Add support for veg_hist forcings (non-climatological)

### LINKED ISSUE #722: Cleanup: miscellaneous cleanup of vic_run
- Description of feature
Clean up miscellaneous unnecessary/messy code in vic_run().
- Description of the problem the feature addresses
vic_run() contains some unnecessary/messy code that could be cleaned up without impacting the physics.  Here's a running list:

- [ ] the local arrays out_prec, out_rain, out_snow, and Melt are defined with twice the length they need, and a factor of 2 is included when indexing them - these are holdovers from the distributed_precip code that no longer exists.
- [ ] the initialization and "preparation" stage before the call to surface_fluxes() is a mix of operations, some of which are only performed once per veg tile and others of which are performed once per veg/band combination.  This leads to a few different loops over bands, as well as a loop over bands stuck inside prepare_full_energy() (see next task).  The order of these tasks appears flexible, so that once-per-veg tile operations could be grouped together outside the loop over bands.  Minimizing the number of loops over bands would shorten the code and make it easier to modify in the future (e.g., the local pointers to the cell, veg_var, etc data structures could be defined inside the loop over bands and we could eliminate a lot of [indexes][indexes]).
- [ ] prepare_full_energy() operates on a single veg tile, but loops over all bands; the entire all_vars struct is passed into it, along with iveg, so that it can single out the veg tile.  Either prepare_full_energy() should be modified to loop over both veg and band dimensions, or loop over neither.
- Will the proposed feature be backward compatible?
Yes.
- Will the proposed feature change the science results of VIC?
No.

### PR DESCRIPTION
**Pull request by tbohn**
- [x] closes #404, #705, #722, #759, #763, #764
- [x] tests passed
- [ ] new tests added
- [ ] science test figures
- [ ] ran uncrustify prior to final commit
- [x] ReleaseNotes entry

### CODE REVIEW

**Comment by jhamman** [comment]
Thanks @tbohn. Can you update your branch with the updates I merged in to `develop` today? I'm going to let @dgergel have the first review here and @bartnijssen and I will follow her up.

**Comment by tbohn** [comment]
Actually, @jhamman @dgergel and @bartnijssen can you hold on a little bit?  The pull request isn't final yet; I may make another simplification.  I will definitely merge the latest develop mods into my branch too.

I'll let you know when this is ready.  Thanks!

**Comment by dgergel** [comment]
@tbohn sure. Just let me know when it's ready.

File: vic/vic_run/include/vic_run.h [hunk 1]
```diff
@@ -186,41 +210,51 @@ int get_volume(lake_con_struct, double, double *);
...
+int newt_raph(void (*vecfunc)(double *, double *, int, int, ...), double *, int);
 double penman(double, double, double, double, double, double, double);
 void photosynth(char, double, double, double, double, double, double, double,
                 double, double, char *, double *, double *, double *, double *,
                 double *);
 void polint(double xa[], double ya[], int n, double x, double *y, double *dy);
-void prepare_full_energy(int, all_vars_struct *, soil_con_struct *, double *,
+void prepare_full_energy(cell_data_struct *, energy_bal_struct *,
+                         soil_con_struct *, double *,
                          double *);
```
Comment (tbohn) [inline vic/vic_run/include/vic_run.h]: The change to prepare_full_energy() is the only non-format modification.  The rest of the changes in vic_run.h came from running uncrustify.

File: docs/Documentation/Drivers/Classic/GlobalParam.md [hunk 2]
```diff
@@ -218,7 +218,7 @@ The following options describe the input parameter files.
...
+| LAI_SRC            | string          | N/A                 | This option tells VIC where to look for LAI values:FROM_VEGLIB = Use the LAI values listed in the vegetation library file.FROM_VEGPARAM = Use the LAI values listed in the vegetation parameter file. Note: for this to work, VEGPARAM_LAI must be TRUE.FROM_VEGHIST = Use the LAI values listed in the veg_hist forcing files. Note: for this to work, LAI must be supplied in the veg_hist files and listd in the global parameter file as one of the variables in the files. Default = FROM_VEGLIB.                                                                                                                                                                 |
```
[the remark this answers is not in the document]
Comment (tbohn) [reply docs/Documentation/Drivers/Classic/GlobalParam.md]: OK, done.

File: docs/Documentation/Drivers/Image/GlobalParam.md [hunk 3]
```diff
@@ -176,7 +176,7 @@ The following options describe the input parameter files.
...
+| LAI_SRC            | string | N/A           | This option tells VIC where to look for LAI values: FROM_VEGLIB = Use the LAI values listed in the vegetation library file.FROM_VEGPARAM = Use the LAI values listed in the vegetation parameter file. Note: for this to work, VEGPARAM_LAI must be TRUE.FROM_VEGHIST = Use the LAI values listed in the veg_hist forcing files. Note: for this to work, LAI must be supplied in the veg_hist files and listd in the global parameter file as one of the variables in the files. Default = FROM_VEGLIB.                                                                                                                                                                |
```
[the remark this answers is not in the document]
Comment (tbohn) [reply docs/Documentation/Drivers/Image/GlobalParam.md]: OK, done.

File: vic/drivers/shared_image/src/vic_init.c [hunk 4]
```diff
@@ -988,11 +989,18 @@ vic_init(void)
...
+                             "equal 1 (%f), dividing each fraction by the "
+                             "sum\n%s", sum, locstr);
+                    for (j = 0; j < options.SNOW_BAND; j++) {
+                        soil_con[i].AreaFract[j] /= sum;
+                    }
+                }
+                else {
+                    log_warn("Sum of the snow band area fractions is 0, "
+                             "setting first fraction to 1\n%s", locstr);
+                    soil_con[i].AreaFract[0] = 1.;
```
[the remark this answers is not in the document]
Comment (tbohn) [reply vic/drivers/shared_image/src/vic_init.c]: OK, done.

**Comment by jhamman** [comment]
@yixinmao and @dgergel - waiting on reviews from each of you on this PR.

**Comment by yixinmao** [comment]
Looks good to me as long as the tests pass. Thanks!

**Comment by dgergel** [comment]
@jhamman the tests are passing on this now, so it should be ready to merge

**Comment by jhamman** [comment]
@tbohn - can you add a note in `ReleaseNotes.md`. Then this will be ready.

**Comment by tbohn** [comment]
I'm a little hesitant to run uncrustify.  Last time I tried, using the config file supplied, the changes were quite extensive and we had to roll them back.  But I'll give it a try locally and see what happens.

**Comment by tbohn** [comment]
Also, what type of comments should I add to ReleaseNotes.md?  Should I describe all the code changes (briefly of course) or just the changes to the user interface (changed name of some variables/options)?

**Comment by jhamman** [comment]
I'll leave the content in the release note up to you. Look at what we wrote leading up to 5.0.

**Comment by tbohn** [comment]
Running `uncrustify` produces many changes, to many files that I didn't even make changes to.  And I'm not thinking you'll like the types of format changes it makes.  For example, from diffing the after and before, we can see:
`diff -r VIC.uncrustify/vic/drivers/cesm/src/vic_cesm_init_library.c VIC/vic/drivers/cesm/src/vic_cesm_init_library.c`
`183c183`
`<         l2x_vic[i].l2x_Fall_lwup = -1 *param.EMISS_GRND *CONST_STEBOL *pow(`
`---`
`>         l2x_vic[i].l2x_Fall_lwup = -1 * param.EMISS_GRND * CONST_STEBOL * pow(`

so, `uncrustify`, at least on my machine, is making multiplication look like pointer dereferencing.

I ran uncrustify via the `run_uncrustify.bash` script.

**Comment by tbohn** [comment]
I've found a potential bug in my changes - nothing major; can fix it by moving a block of code.  But please wait before merging this PR until I've properly fixed the problem.

**Comment by jhamman** [comment]
@tbohn - is this ready to go?

**Comment by tbohn** [comment]
yes, this is ready.

**Comment by jhamman** [comment]
@yixinmao - can you look into why these tests are failing now? Is it a Python library version issue?

**Comment by yixinmao** [comment]
@tbohn : I briefly looked at the test results, and it seems like for the classic driver, there are some `nan`'s in the flux output files; and with `FULL_ENERGY=TRUE` and `FROZEN_SOIL=TRUE`, it seems like there are some issue with iteration convergence so that the run doesn't end (at least not for a long time). These issues occur locally (so not a travis thing). Do you have any idea what are the causes?

**Comment by tbohn** [comment]
@yixinmao : strange, no, I don't know what is causing those NaNs.  The only things I did were to (a) merge in the latest develop changes and (b) move a block of code to a different place in vic_run.c.  It is, of course, possible that I caused the NaNs by doing so, although in my tests on my own machine those didn't occur.  I'll take a look.

**Comment by yixinmao** [comment]
@tbohn hmm weird. Did all tests pass on your machine? They do fail on mine the same way that travis tests fail (i.e., system test failure as well as  out of time). Let me know what you find (or can't find).

**Comment by tbohn** [comment]
@yixinmao I wasn't referring to travis tests on my system; I was referring to tests at my own test sites.  Another shortcoming of my tests, which is completely my fault, is that I was using a copy of my git repository for dev/testing, and then ported the changes into git.  So there's a chance that the testing version of the code was fine but my port somehow messed things up.

**Comment by yixinmao** [comment]
@tbohn OK. The tests failed on my local machine without travis (I fetched your branch directly). Do you think you could take a look to see what might be going on? It's probably faster than me looking from scratch. But if you still can't find anything, let me know and I can take a closer look.

**Comment by tbohn** [comment]
@yixinmao yes, I will take a look this afternoon - I don't expect you to do that investigation; it's my responsibility to debug it, I think.  I only needed help earlier in understanding the nature of the travis errors (the messages were not clear to me).

**Comment by yixinmao** [comment]
@jhamman : after @tbohn fixes all tests passed and this PR is ready to be merged.

**Comment by jhamman** [comment]
This looks good to me but I'd like @bartnijssen to give it a final look before merging.

**Pull request merged**